File: dsa-problems/Trees/solutions/python/TRE-013-auditorium-bst-validate-gap.py
import sys

# Increase recursion depth
sys.setrecursionlimit(200000)

def validate_bst_gap(n: int, values: list[int], left: list[int], right: list[int], G: int) -> bool:
    if n == 0:
        return True
        
    def validate(u, min_val, max_val):
        if u == -1:
            return True
            
        val = values[u]
        # BST Check
        if val <= min_val or val >= max_val:
            return False
            
        # Left Child
        if left[u] != -1:
            l_val = values[left[u]]
            if abs(val - l_val) < G:
                return False
            if not validate(left[u], min_val, val):
                return False
                
        # Right Child
        if right[u] != -1:
            r_val = values[right[u]]
            if abs(val - r_val) < G:
                return False
            if not validate(right[u], val, max_val):
                return False
                
        return True

    return validate(0, float('-inf'), float('inf'))

def main():
    data = sys.stdin.read().strip().split()
    if not data:
        return
    idx = 0
    n = int(data[idx]); idx += 1
    values = [0] * n
    left = [0] * n
    right = [0] * n
    for i in range(n):
        values[i] = int(data[idx]); idx += 1
        left[i] = int(data[idx]); idx += 1
        right[i] = int(data[idx]); idx += 1
    G = int(data[idx]) if idx < len(data) else 0
    
    print("true" if validate_bst_gap(n, values, left, right, G) else "false")

if __name__ == "__main__":
    main()

File: dsa-problems/Trees/solutions/java/TRE-013-auditorium-bst-validate-gap.java
import java.util.*;

class Solution {
    public boolean validateBSTGap(int n, long[] values, int[] left, int[] right, long G) {
        if (n == 0) return true;
        return validate(0, Long.MIN_VALUE, Long.MAX_VALUE, values, left, right, G);
    }

    private boolean validate(int u, long min, long max, long[] values, int[] left, int[] right, long G) {
        if (u == -1) return true;

        long val = values[u];
        // BST Check
        if (val <= min || val >= max) return false;

        // Left Child Check
        if (left[u] != -1) {
            long lVal = values[left[u]];
            if (Math.abs(val - lVal) < G) return false; // Gap Check
            if (!validate(left[u], min, val, values, left, right, G)) return false;
        }

        // Right Child Check
        if (right[u] != -1) {
            long rVal = values[right[u]];
            if (Math.abs(val - rVal) < G) return false; // Gap Check
            if (!validate(right[u], val, max, values, left, right, G)) return false;
        }

        return true;
    }
}

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt()) return;
        int n = sc.nextInt();
        long[] values = new long[n];
        int[] left = new int[n];
        int[] right = new int[n];
        for (int i = 0; i < n; i++) {
            values[i] = sc.nextLong();
            left[i] = sc.nextInt();
            right[i] = sc.nextInt();
        }
        long G = 0;
        if (sc.hasNextLong()) G = sc.nextLong();

        Solution solution = new Solution();
        System.out.println(solution.validateBSTGap(n, values, left, right, G) ? "true" : "false");
        sc.close();
    }
}
